File: xdef.net/xdef.net/Connection/Library/RemoteStreamWrapper.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using xdef.net.Utils;

namespace xdef.net.Connection.Library
{
    internal class RemoteStreamWrapper : RemoteHandlingObject
    {
        private const int FUNCTION_AVAILABLE = 1;
        private const int FUNCTION_CLOSE = 2;
        private const int FUNCTION_READ = 3;
        private const int FUNCTION_WRITE = 4;

        private readonly Stream _stream;

        public RemoteStreamWrapper(Client client, Stream stream) : base(client)
        {
            _stream = stream;
        }

        public override Request HandleRequest(Request request)
        {
            switch (request.Function)
            {
                case FUNCTION_AVAILABLE:
                    return Available();
                case FUNCTION_CLOSE:
                    return Close();
                case FUNCTION_READ:
                    return Read(request);
                case FUNCTION_WRITE:
                    return Write(request);
                default:
                    return new ResponseException(Response.ERROR_CODE_UNKNOWN_FUNCTION, $"Unknown function {request.Function} on {nameof(RemoteStreamWrapper)}");
            }
        }

        private Request Write(Request request)
        {
            using (var reader = request.Reader)
            {
                var len = reader.ReadInt32();
                var buf = reader.ReadBytes(len);
                _stream.Write(buf, 0, len);
            }
            return null;
        }

        private Request Read(Request request)
        {
            using (var reader = request.Reader)
            {
                var len = reader.ReadInt32();
                var buf = new byte[len];
                var bytesRead = _stream.Read(buf, 0, len);
                using (var builder = new BigEndianDataBuilder())
                {
                    builder.Add(bytesRead)
                        .Add(buf);
                    return new Response(builder.Build());
                }
            }
        }

        private Request Close()
        {
            _stream.Close();
            return null;
        }

        private Request Available()
        {
            var avail = (int)(_stream.Length - _stream.Position);
            return new Response(BigEndianBitConverter.GetBytes(avail));
        }

    }
}

File: xdef.bridge/xdef.bridge/src/main/java/org/xdef/bridge/wrappers/streams/RemoteStreamWrapper.java
package org.xdef.bridge.wrappers.streams;

import java.io.IOException;

import org.xdef.bridge.remoteObjects.RemoteObject;
import org.xdef.bridge.server.Client;
import org.xdef.bridge.server.requests.Request;
import org.xdef.bridge.server.requests.Response;
import org.xdef.bridge.utils.BinaryDataBuilder;
import org.xdef.bridge.utils.BinaryDataReader;

public class RemoteStreamWrapper extends RemoteObject {

    private static final int FUNCTION_AVAILABLE = 1;
    private static final int FUNCTION_CLOSE = 2;
    private static final int FUNCTION_READ = 3;
    private static final int FUNCTION_WRITE = 4;

    public RemoteStreamWrapper(Client client, int objectId) {
        super(client, objectId);
    }

    public int available() {
        Request response = sendRequestWithResponse(new Request(FUNCTION_AVAILABLE, null));
        try {
            return response.getReader().readInt();
        } catch (IOException e) {
            return 0;
        }
    }

    public void close() {
        sendRequest(new Request(FUNCTION_CLOSE, null));
    }

    public synchronized int read(byte[] buf, int offset, int len) {
        BinaryDataBuilder builder = new BinaryDataBuilder()
        .add(len);
        Request response = sendRequestWithResponse(new Request(FUNCTION_READ, builder.build()));
        try {
            BinaryDataReader reader = response.getReader();
            int payload = reader.readInt();
            byte[] data = reader.readNBytes(payload);
            System.arraycopy(data, 0, buf, offset, payload);
            return payload;
        } catch (IOException ex) {
            return 0;
        }
    }

    public synchronized void write(byte[] buf, int offset, int len) {
        BinaryDataBuilder builder = new BinaryDataBuilder()
                .add(len)
                .add(buf, offset, len);
        sendRequest(new Request(FUNCTION_WRITE, builder.build()));
    }

}